domain/src/main/java/org/oppia/domain/topic/StoryProgressController.kt

```diff
@@ -0,0 +1,165 @@
+  //  indicates that they can't be started due to previous chapter not yet being completed.
+
+  private val trackedStoriesProgress: Map<String, TrackedStoryProgress> by lazy { createInitialStoryProgressState() }
+
+  /**
+   * Records the specified chapter completed within the context of the specified story. Returns a [LiveData] that
+   * provides exactly one [AsyncResult] to indicate whether this operation has succeeded. This method will never return
+   * a pending result.
+   */
+  fun recordCompletedChapter(storyId: String, explorationId: String): LiveData<AsyncResult<Nothing?>> {
+    return try {
+      trackCompletedChapter(storyId, explorationId)
+      MutableLiveData(AsyncResult.success(null))
+    } catch (e: Exception) {
+      MutableLiveData(AsyncResult.failed(e))
+    }
+  }
+
+  // TODO(#21): Implement notifying story progress changes when a chapter is recorded as complete, and add tests for
+  //  this case.
+
+  /**
+   * Returns a [LiveData] corresponding to the story progress of the specified story, or a failure if no such story can
+   * be identified. This [LiveData] will update as the story's progress changes.
+   */
+  fun getStoryProgress(storyId: String): LiveData<AsyncResult<StoryProgress>> {
+    return try {
+      MutableLiveData(AsyncResult.success(createStoryProgressSnapshot(storyId)))
+    } catch (e: Exception) {
+      MutableLiveData(AsyncResult.failed(e))
+    }
+  }
+
+  private fun trackCompletedChapter(storyId: String, explorationId: String) {
+    check(storyId in trackedStoriesProgress) { "No story found with ID: $storyId" }
+    trackedStoriesProgress.getValue(storyId).markChapterCompleted(explorationId)
+  }
+
+  private fun createStoryProgressSnapshot(storyId: String): StoryProgress {
+    check(storyId in trackedStoriesProgress) { "No story found with ID: $storyId" }
+    return trackedStoriesProgress.getValue(storyId).toStoryProgress()
+  }
+
+  private fun createInitialStoryProgressState(): Map<String, TrackedStoryProgress> {
+    return mapOf(
+      TEST_STORY_ID_0 to createStoryProgress0(),
+      TEST_STORY_ID_1 to createStoryProgress1(),
+      TEST_STORY_ID_2 to createStoryProgress2()
+    )
+  }
+
+  private fun createStoryProgress0(): TrackedStoryProgress {
+    return TrackedStoryProgress(
+      chapterList = listOf(TEST_EXPLORATION_ID_0),
+      completedChapters = setOf(TEST_EXPLORATION_ID_0)
+    )
+  }
+
+  private fun createStoryProgress1(): TrackedStoryProgress {
+    return TrackedStoryProgress(
+      chapterList = listOf(TEST_EXPLORATION_ID_1, TEST_EXPLORATION_ID_2, TEST_EXPLORATION_ID_3),
+      completedChapters = setOf(TEST_EXPLORATION_ID_1)
+    )
+  }
+
+  private fun createStoryProgress2(): TrackedStoryProgress {
+    return TrackedStoryProgress(
+      chapterList = listOf(TEST_EXPLORATION_ID_4),
+      completedChapters = setOf()
+    )
+  }
+
+  private class TrackedStoryProgress(private val chapterList: List<String>, completedChapters: Set<String>) {
```

Comment:
Can we add comments for this class to describe this to everyone?

Reply:
Sure, done.